[SPARK-20384][SQL] Support value class in schema of Dataset (without breaking existing current projection)

mickjermsurawong-stripe opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
- This PR revisits https://github.com/apache/spark/pull/22309, and [SPARK-20384](https://issues.apache.org/jira/browse/SPARK-20384) solving the original problem, but additionally will prevent backward-compat break on schema of top-level `AnyVal` value class.
- Why previous break? We currently support top-level value classes just as any other case class; field of the underlying type is present in schema. This means any dataframe SQL filtering on this expects the field name to be present. The previous PR changes this schema and would result in breaking current usage. See test `"schema for case class that is a value class"`. This PR keeps the schema.
- We actually currently support collection of value classes prior to this change, but not case class of nested value class. This means the schema of these classes shouldn't change to prevent breaking too. See the tests asserting schema before any changes in 0cdad3b
- However, what we can change, without breaking, is schema of nested value class, which will fails due to the compile problem, and thus its schema now isn't actually valid. After the change, the schema of this nested value class is now flattened: 
c7aaae8
- With this PR, there's flattening only for nested value class (new), but not for top-level and collection classes (existing behavior)


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
- Currently, nested value class isn't supported. This is because when the generated code treats `anyVal` class in its unwrapped form, but we encode the type to be the wrapped case class. This results in compile of generated code
For example, 
For a given `AnyVal` wrapper and its root-level class container
```
case class IntWrapper(i: Int) extends AnyVal
case class ComplexValueClassContainer(c: IntWrapper)
```
The problematic part of generated code:
```
    private InternalRow If_1(InternalRow i) {
        boolean isNull_42 = i.isNullAt(0);
        // 1) ******** The root-level case class we care
        org.apache.spark.sql.catalyst.encoders.ComplexValueClassContainer value_46 = isNull_42 ?
            null : ((org.apache.spark.sql.catalyst.encoders.ComplexValueClassContainer) i.get(0, null));
        if (isNull_42) {
            throw new NullPointerException(((java.lang.String) references[5] /* errMsg */ ));
        }
        boolean isNull_39 = true;
        // 2) ******** We specify its member to be unwrapped case class extending `AnyVal`
        org.apache.spark.sql.catalyst.encoders.IntWrapper value_43 = null;
        if (!false) {

            isNull_39 = false;
            if (!isNull_39) {
                // 3) ******** ERROR: `c()` compiled however is of type `int` and thus we see error
                value_43 = value_46.c();
            }
        }
```
We get this errror: Assignment conversion not possible from type "int" to type "org.apache.spark.sql.catalyst.encoders.IntWrapper"
```
java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: 
File 'generated.java', Line 159, Column 1: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 159, Column 1: Assignment conversion not possible from type "int" to type "org.apache.spark.sql.catalyst.encoders.IntWrapper"
```

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
- Yes, this will fix the bug.


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
- This PR ports all the tests from https://github.com/apache/spark/pull/22309 and added a few more.

cc @mt40 please let me know if i'm missing something from your previous PR
cc @joshrosen-stripe

mickjermsurawong-stripe committed:
failing test on nested/complex value

mickjermsurawong-stripe committed:
assert current spark schema for collection of values

mickjermsurawong-stripe committed:
unwrap params passing encoder suite

clean up

mickjermsurawong-stripe committed:
fix test: inner value class unwrapped

mickjermsurawong-stripe committed:
clean up case class instantiation

mickjermsurawong-stripe committed:
add more test for nested collection

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

JoshRosen commented:
jenkins this is ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21206/
Test PASSed.

SparkQA commented:
**[Test build #116417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116417/testReport)** for PR 27153 at commit [`6cb83b4`](https://github.com/apache/spark/commit/6cb83b4ab9f2428092c4323c887f31e6dd58aaf6).

SparkQA commented:
**[Test build #116417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116417/testReport)** for PR 27153 at commit [`6cb83b4`](https://github.com/apache/spark/commit/6cb83b4ab9f2428092c4323c887f31e6dd58aaf6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SeqOfValueClass(s: Seq[StringWrapper])`
  * `case class MapOfValueClassKey(m: Map[IntWrapper, String])`
  * `case class MapOfValueClassValue(m: Map[String, StringWrapper])`
  * `case class OptionOfValueClassValue(o: Option[StringWrapper])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116417/
Test PASSed.

mt40 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -753,6 +753,24 @@ object ScalaReflection extends ScalaReflection {
+   * @return unwrapped param
+   */
+  private def unwrapValueClassParam(param: (String, `Type`)): (String, `Type`) = {
+    val (name, typ) = param
```
nit: maybe use `tpe` instead of `typ` to be consistent with the naming in this file

mt40 commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala:
```diff
@@ -146,8 +146,20 @@ object TraitProductWithNoConstructorCompanion {}
+  case class StrWrapper(s: String) extends AnyVal
+
+  case class ValueClassData(
+                             intField: Int,
```
please fix indentation here

mt40 commented:
@mickjermsurawong-stripe thanks for continue with this feature, I hope it will be merged this time

mickjermsurawong-stripe committed:
fix nits

mickjermsurawong-stripe committed:
add filter tests for explicitness in sql schema

SparkQA commented:
**[Test build #116688 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116688/testReport)** for PR 27153 at commit [`c23705d`](https://github.com/apache/spark/commit/c23705d9715efa16d73491191143c9f802bdff51).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21467/
Test PASSed.

mickjermsurawong-stripe commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -679,6 +679,33 @@ class DataFrameSuite extends QueryTest with SharedSparkSession {
+    val df = spark.sparkContext.parallelize(Seq(a, b)).toDF
+    // flat value class, `s` field is not in schema
+    val filtered = df.where("wrapper = \"a\"")
+    checkAnswer(filtered, spark.sparkContext.parallelize(Seq(a)).toDF)
```
Before this change we never support nested value class: 
- Filter with `wrapper` would break with
```
org.apache.spark.sql.AnalysisException: cannot resolve '(`wrapper` = 'a')' due to data type mismatch: differing types in '(`wrapper` = 'a')' (struct<s:string> and string).; line 1 pos 0;
```
- Filter with `wrapper.s` would break with:
```
java.lang.ClassCastException: java.lang.String cannot be cast to org.apache.spark.sql.test.SQLTestData$StringWrapper
```

mickjermsurawong-stripe commented:
Thanks @mt40!
Also added more tests on filtering on dataframe c23705d to illustrate the expected schemas more explicitly.

SparkQA commented:
**[Test build #116688 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116688/testReport)** for PR 27153 at commit [`c23705d`](https://github.com/apache/spark/commit/c23705d9715efa16d73491191143c9f802bdff51).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  test(\"Value class filter\") `
  * `  test(\"Array value class filter\") `
  * `  test(\"Nested value class filter\") `
  * `  case class StringWrapper(s: String) extends AnyVal`
  * `  case class ArrayStringWrapper(wrappers: Seq[StringWrapper])`
  * `  case class ContainerStringWrapper(wrapper: StringWrapper)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116688/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #117086 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117086/testReport)** for PR 27153 at commit [`c23705d`](https://github.com/apache/spark/commit/c23705d9715efa16d73491191143c9f802bdff51).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21850/
Test PASSed.

SparkQA commented:
**[Test build #117086 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117086/testReport)** for PR 27153 at commit [`c23705d`](https://github.com/apache/spark/commit/c23705d9715efa16d73491191143c9f802bdff51).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  test(\"Value class filter\") `
  * `  test(\"Array value class filter\") `
  * `  test(\"Nested value class filter\") `
  * `  case class StringWrapper(s: String) extends AnyVal`
  * `  case class ArrayStringWrapper(wrappers: Seq[StringWrapper])`
  * `  case class ContainerStringWrapper(wrapper: StringWrapper)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117086/
Test FAILed.

mickjermsurawong-stripe commented:
^ bumping this please. I suspect the patch failure here is flaky; change from the last build is only adding more tests.

eejbyfeldt commented:
I am interested in support for value classes that is added/fixed in this branch. To me the changes looks like they are a still valid approach of adding the support. 

Rebasing this branch on master while cause some of the added test cases to fail. This is due to a PR https://github.com/apache/spark/pull/31766 the fixed the so that interpreted path test are run properly and a bug for
CatalystToExternalMap in the interpreted path. I created a PR here to address that bug here: https://github.com/apache/spark/pull/32783

If/after the bug fix gets merged, would the next step towards getting this merged? Should I open a new PR with the changes from @mickjermsurawong-stripe branch? Since to me it looks like the original author is no longer around and/or interested in these changes.

mt40 commented:
@eejbyfeldt Nice to see someone having the same interest in this. I think you should continue working on this branch to avoid duplication but it's up to you though.

In my opinion, the hardest part would be to convince Spark team that this change is necessary so this is can be merged once failed tests are fixed.

mickjermsurawong-stripe commented:
hi @eejbyfeldt! We've had this equivalent change in our forked running in prod for a while now. One more follow-up we had internally is to handle value class wrapped in tuple like `Dataset[(MyAnyValKey, SomeOtherCaseClass)]`. I could revive this PR and add the said patch in the next few days.

eejbyfeldt commented:
Hi @mickjermsurawong-stripe ! Great to hear, looking forward to updated changes. Also my PR https://github.com/apache/spark/pull/32783 got merged, so rebasing this branch on master should no longer fail tests.

eejbyfeldt commented:
@mickjermsurawong-stripe Have you had anytime to look at updating the branch? I you don't have time to prepare full patch, I could help out with unittests and such if you point can provide some information about approximatly which change is needed for the value classes in tuples.

eejbyfeldt commented:
I looked some more at this and realised what the problem with `Tuples` is, it guess it all related to when a value class it its underlying type and when it is not https://docs.scala-lang.org/overviews/core/value-classes.html#allocation-details

Based on this it seems something like the approach of the PR https://github.com/apache/spark/pull/22309 by @mt40 would be needed to handle all cases correctly. To me it also makes sense that a AnyVal would give the same schema regardless of where it is used like that PR. So unless there is someother good suggestion on this, that is the approach I am going to persue.

mickjermsurawong-stripe commented:
Hey @eejbyfeldt  sorry i realized that the internal fork we addressed this issue is in a different class of our custom encoder, so i didn't to integrate it with the standard encoder here.

The underlying issue we addressed is exactly the allocation you mentioned: From doc on value class: https://docs.scala-lang.org/overviews/core/value-classes.html Given: `class Wrapper(val underlying: Int) extends AnyVal`,

- "The type at compile time is Wrapper, but at runtime, the representation is an Int"
This implies that when our struct has a field of AnyVal case class, our generated code
should support the underlying type during runtime execution.

- The Wrapper "must be instantiated... when a value class is used as a type argument".
This implies that scala.Tuple[Wrapper, ...], Seq[Wrapper], Map[String, Wrapper], Option[Wrapper] will still contain Wrapper as-is in during runtime instead of Int.

Hope this could help. It's a long weekend here, so I will make sure I'll get to work on this. But please feel free to pursue independently as well.

mickjermsurawong-stripe commented:
hi @eejbyfeldt i made another PR to address the tuple issue raised. https://github.com/apache/spark/pull/33205. 
Would appreciate your review as well.
